auth/common.cc

```diff
@@ -202,4 +203,85 @@ future<> announce_mutations(
     co_await announce_mutations_with_guard(group0_client, std::move(cmuts), std::move(group0_guard), as);
 }
 
+mutations_collector::~mutations_collector() {
+    if (_muts.size() > 0 || _generators.size() > 0) {
+        on_internal_error_noexcept(auth_log, format("mutations_collector: contains {} mutations and {} generators when destructed, they will be lost", _muts.size(), _generators.size()));
+    }
+}
+
+api::timestamp_type mutations_collector::write_timestamp() const {
+    return _guard->write_timestamp();
+}
+
+void mutations_collector::add_mutation(mutation m) {
+    _muts.push_back(std::move(m));
+}
+
+void mutations_collector::add_mutations(std::vector<mutation> ms) {
+    _muts.insert(_muts.end(),
+            std::make_move_iterator(ms.begin()),
+            std::make_move_iterator(ms.end()));
+}
+
+void mutations_collector::add_generator(generator_func f) {
+    _generators.push_back(std::move(f));
+}
+
+future<> mutations_collector::announce(::service::raft_group0_client& group0_client, seastar::abort_source& as) {
+    if (_muts.size() == 0 && _generators.size() == 0) {
+        co_return;
+    }
+    if (!_guard) {
+        on_internal_error(auth_log, "mutations_collector: trying to announce without guard");
+    }
+    // common case, don't bother with potential batching as typically we
+    // would have only 1-2 mutations, when producer expects higher number of
+    // mutations it should use generator
+    if (_generators.size() == 0) {
+        std::vector<canonical_mutation> cmuts = {_muts.begin(), _muts.end()};
+        _muts.clear();
+        co_return co_await announce_mutations_with_guard(group0_client, std::move(cmuts), std::move(*_guard), &as);
+    }
+
+    if (_muts.size() > 0) {
+        _generators.emplace_back([this] (api::timestamp_type&) -> mutations_generator {
+            for (auto& mut: _muts) {
+                co_yield std::move(mut);
+            }
+            _muts.clear();
+        });
+    }
+
+    auto chained_generator = [this] (api::timestamp_type& t) -> mutations_generator {
+        for (auto& generator: _generators) {
+            auto g = generator(t);
+            while (auto mut = co_await g()) {
+                co_yield std::move(*mut);
+            }
+        }
+        _generators.clear();
+    };
+
+    co_return co_await announce_mutations_with_batching(
+            group0_client,
+            std::bind_front(&::service::raft_group0_client::start_operation, &group0_client),
```

Comment:
Don't you need to drop '_guard' since the function will try to take another one.

Reply:
yes, I need to refactor this further